BridgeAppSDK/SBAUserWrapper+Bridge.swift

```diff
@@ -0,0 +1,311 @@
+     */
+    public func loginUser(externalId: String, completion: ((NSError?) -> Void)?) {
+        let (email, password) = emailAndPasswordForExternalId(externalId)
+        guard (email != nil) && (password != nil) else {
+            return
+        }
+        signInUser(email!, password: password!, completion: completion)
+    }
+    
+    /**
+     * Login a user on this device who has previously completed registration on a different device.
+     */
+    public func loginUser(email email: String, password: String, completion: ((NSError?) -> Void)?) {
+        signInUser(email, password: password) { [weak self] (error) in
+            guard (self != nil) else { return }
+            
+            if ((error == nil) || error!.code == SBBErrorCode.ServerPreconditionNotMet.rawValue) {
+                self!.email = email
+                self!.password = password
+            }
+            self!.callCompletionOnMain(error, completion: completion)
+        }
+    }
+
+    /**
+     * Send user consent signature (if server precondition not met and reconsenting user)
+     */
+    public func sendUserConsented(consentSignature: SBAConsentSignature, completion: ((NSError?) -> Void)?) {
+        
+        let name = consentSignature.signatureName ?? self.name ?? "First Last"
+        let birthdate = consentSignature.signatureBirthdate?.startOfDay() ?? NSDate(timeIntervalSince1970: 0)
+        let consentImage = consentSignature.signatureImage
+        let subpopGuid = self.subpopulationGuid ?? SBAAppDelegate.sharedDelegate?.bridgeInfo.studyIdentifier ?? "unknown"
+        
+        SBAUserBridgeManager.sendUserConsented(name, birthDate: birthdate, consentImage: consentImage, sharingScope: self.dataSharingScope, subpopulationGuid: subpopGuid) { [weak self] (_, error) in
+            guard (self != nil) else { return }
+            
+            self!.consentVerified = (error == nil)
+            self!.callCompletionOnMain(error, completion: completion)
+        }
+    }
+
+    /**
+     * Sign in when app is active if the login and consent have been verified
+     */
+    public func ensureSignedInWithCompletion(completion: ((NSError?) -> Void)?) {
+        
+        // If the user is not logged in or consented then do not attempt login
+        // Just return with an error
+        guard self.loginVerified && self.consentVerified else {
+            let code = !self.loginVerified ? SBBErrorCode.NoCredentialsAvailable.rawValue : SBBErrorCode.ServerPreconditionNotMet.rawValue
+            let error = NSError(domain: SBAUserErrorDomain, code: code, userInfo: nil)
+            self.callCompletionOnMain(error, completion: completion)
+            return
+        }
+        
+        // Make sure that 
+        SBAUserBridgeManager.ensureSignedInWithCompletion { [weak self] (_, error) in
+            guard (self != nil) else { return }
+            
+            if (error != nil) && error!.code == SBBErrorCode.ServerPreconditionNotMet.rawValue {
```

Comment:
FYI if you're already signed in (i.e. have a session token), no attempt will be made to sign you in again and error will be nil here. You'll still get the 412 on the next attempt to call an API that requires consent, and it'll be handled then, but as I said, just FYI in case it wasn't obvious.